src/dbnode/storage/bootstrap/types.go

```diff
@@ -138,82 +224,44 @@ type BootstrapperProvider interface {
-	) (result.DataBootstrapResult, error)
-
-	// BootstrapIndex performs bootstrapping of index blocks for the given time ranges, returning
-	// the bootstrapped index blocks and the time ranges it's unable to fulfill in parallel. A bootstrapper
-	// should only return an error should it want to entirely cancel the bootstrapping of the
-	// node, i.e. non-recoverable situation like not being able to read from the filesystem.
-	BootstrapIndex(
-		ns namespace.Metadata,
-		shardsTimeRanges result.ShardTimeRanges,
-		opts RunOptions,
-	) (result.IndexBootstrapResult, error)
+	// Bootstrap performs bootstrapping of series data and index metadata
+	// for the  given time ranges, returning the bootstrapped series data
+	// and index blocks for the time ranges it's unable to fulfill in parallel.
+	// A bootstrapper should only return an error should it want to entirely
+	// cancel the bootstrapping of the node, i.e. non-recoverable situation
+	// like not being able to read from the filesystem.
+	Bootstrap(namespaces Namespaces) (NamespaceResults, error)
 }
 
 // Source represents a bootstrap source. Note that a source can and will be reused so
 // it is important to not rely on state stored in the source itself with the mindset
 // that it will always be set to default values from the constructor.
 type Source interface {
-	// Can returns whether a specific bootstrapper strategy can be applied.
-	Can(strategy Strategy) bool
-
 	// AvailableData returns what time ranges are available for bootstrapping a given set of shards.
 	AvailableData(
 		ns namespace.Metadata,
 		shardsTimeRanges result.ShardTimeRanges,
 		runOpts RunOptions,
 	) (result.ShardTimeRanges, error)
 
-	// ReadData returns raw series for a given set of shards & specified time ranges and
-	// the time ranges it's unable to fulfill. A bootstrapper source should only return
-	// an error should it want to entirely cancel the bootstrapping of the node,
-	// i.e. non-recoverable situation like not being able to read from the filesystem.
-	ReadData(
-		ns namespace.Metadata,
-		shardsTimeRanges result.ShardTimeRanges,
-		runOpts RunOptions,
-	) (result.DataBootstrapResult, error)
-
 	// AvailableIndex returns what time ranges are available for bootstrapping.
 	AvailableIndex(
 		ns namespace.Metadata,
 		shardsTimeRanges result.ShardTimeRanges,
 		opts RunOptions,
 	) (result.ShardTimeRanges, error)
 
-	// ReadIndex returns series index blocks.
-	ReadIndex(
-		ns namespace.Metadata,
-		shardsTimeRanges result.ShardTimeRanges,
-		opts RunOptions,
-	) (result.IndexBootstrapResult, error)
+	// Read returns series data and index metadata for a given set of shards
+	// and specified time ranges and the time ranges it's unable to fulfill.
+	// A bootstrapper source should only return an error should it want to
+	// entirely cancel the bootstrapping of the node, i.e. non-recoverable
+	// situation like not being able to read from the filesystem.
+	Read(namespaces Namespaces) (NamespaceResults, error)
```

Comment:
Might be a good idea to filter a `NamespaceResultsMapOptions` through here in case this becomes more customizable later? Although that may be over-engineering it

Reply:
Also any way we can use a single `bootstrap.NamespaceResults` and filter it through various `Reads` to avoid recreating the map per bootstrapper? If I'm understanding this right, every successive bootstrap method should have smaller and smaller result maps generated, but should still give us decent savings if we only use one right?

Reply:
So the results map is super tiny thankfully, it literally only has per namespace options.  I'm not too worried about maps of size 5-10 max.  

We removed all the big per series results map by removing the data results type and using the accumulator.